kotest-assertions/src/commonMain/kotlin/io/kotest/properties/shrinking/StringShrinker.kt

```diff
@@ -1,14 +1,22 @@
 package io.kotest.properties.shrinking
 
-object StringShrinker : Shrinker<String> {
-  override fun shrink(failure: String): List<String> = when (failure.length) {
-    0 -> emptyList()
-    1 -> listOf("", "a")
-    else -> {
-      val first = failure.take(failure.length / 2 + failure.length % 2)
-      val second = failure.takeLast(failure.length / 2)
-      // always include empty string as the best io.kotest.properties.shrinking.shrink
-     listOf("", first, first.padEnd(failure.length, 'a'), second, second.padStart(failure.length, 'a'))
-    }
-  }
+class StringShrinker(private val minSize: Int, private val padChar: Char) : Shrinker<String> {
```

Comment:
If anyone uses `StringShrinker`, this code will break theirs. Do we really want to do that?

Reply:
I am not convinced it should even be public. What is the use case where someone needs access to it? To me it seems like simply having access to a customizable `Gen<String>` via `Gen.string()` would be enough for almost any use case. In in mind, if someone is doing something that `Gen.string()`, via the customizable parameters, cannot do then a generic `StringShrinker` would not work for them anyway. I would love to make this class private, and move the tests to `GenStringTests`.

As far as backwards compatibility, I had 4.0 in mind for this code.

@sksamuel What are your thoughts?

Reply:
I've not been following this thread as I've been super busy. What's the gist and how does it fit in with what I was saying in #kotlintest-devs ?

Reply:
@sksamuel 

Overall, for this PR, `Gen.string()` has been rewritten to allow passing in a `Gen<Char>`. This also has implications for the `StringShrinker`.

`StringShrinker` has been changed to respect the min size of a generated `String` and the pad char comes from the passed-in `Gen<Char>`. Some constructor parameters were added to enabled to shrinking to align with the content that the `Gen<String>` creates. I have broken backwards compatibility here by changing `StringShrinker` from an object to a class and requiring constructor parameters there.

My opinion is that `StringShrinker` is not really useful outside the scope of our implementation of `Gen.string()`, that if someone has a custom `Gen<String>` that can do something that `Gen.string()` cannot, then `StringShrinker` would not work for them anyway, as they would need shrinking that is specific to the rules of their custom `Gen<String>`.

Therefore, I am proposing that for 4.0, that we make `StringShrinker` private.

Reply:
I agree.

Reply:
Cool thanks. I will make the changes in this PR.

Reply:
Could we roll out these changes to the new property test code ?